### PR TITLE
Contact Support: Submit ticket to Zendesk with chat history

### LINKED ISSUE #21278: Contact Support: Submit ticket to Zendesk with chat history

### PR DESCRIPTION
**Pull request by staskus**
Fixes #21278,  #21277 

**Progress:**
- [x] Handle support callback from the chat bot and capture chat history
- [x] Integrate with https://github.com/wordpress-mobile/WordPress-iOS/pull/21270 once merged and rebase
- [x] Submit ticket to Zendesk
- [x] Add proper message subject and history formatting
- [x] Add loading indicators, error handling
- [x] Prepare PR for testing

## Solution

Making changes based on [DocsBot.ai docs](https://docsbot.ai/docs/embeddable-chat-widget):
- Adding a `supportLink: "#"` so the button would show up
- Adding a `supportCallback` to pass a history to Swift

Using Zendesk SDK APIs to create a new request, passing formatted message history as a description

## To test:

Temporary: Add id in `SupportChatViewModel`

1. Open Jetpack app
2. Enable "Contact Support via DocsBot" feature flag
3. Help & Support -> Contact Support
4. Ask bot a question
5. Confirm that below the answer "Contact support" button appears
6. Tap "Contact support"
7. Confirm if support identity hasn't been set, alert prompting for email appears
8. Confirm loading indicator is shown
9. Confirm success message is shown
10. Confirm ticket appears under "Tickets" section

## Regression Notes
1. Potential unintended areas of impact

Sending incorrect Zendesk reports

2 What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing

3. What automated tests I added (or what prevented me from doing so)

Unit tests for asserting how we format history and tags for Zendesk

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

Not in the scope of this PR

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by staskus** [comment]
@guarani just letting you know that the work can be continued if you're available, we can also review & merge and continue the work in another ticket

More context: p1691606402570209-slack-C05ALSVN5K5

**Comment by guarani** [comment]
>  Add proper message subject and history formatting

The ticket subject in Zendesk is `Jetpack for iOS Support`, which looks to be the same for all tickets users create from this app. Is there anything we should do here?
About the history formatting:
- I hoped to label each message as from the user or the bot. That would make it easy for support folk to read the transcript once if the user chooses to create a ticket. However, the chat history appears to be a simple array of strings and we can't reliably tell which message is from the user vs. the chatbot. Do you have any thoughts on how to improve this @staskus? 
- To create an empty line between chat messages, I tried `\n\n` but it seems Zendesk ignores these in their backend dashboard (they are rendered when viewing the ticket transcript in My Tickets, as shown in the video below). Zendesk's ticket dashboard might interpret markdown, so another approach could be to do what's suggested [here](https://stackoverflow.com/a/64164898/1305067) and use backslashes, but I haven't tried it yet.

> Add loading indicators, error handling

I think we could use the `Saving..` HUD-like loading indicator shown when saving a draft post, but I haven't tried this out yet. Regarding error handling, I added an error toast message if the ticket creation failed, but haven't tested it yet.

---

Here's how this is looking with the latest changes (shows toast when ticket is created and offers to take user to newly created ticket):

https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/43587437-a1ab-45d8-8b51-a41257f1c720

**Comment by staskus** [comment]
@guarani 

Thank you for taking over and making the improvements!

> The ticket subject in Zendesk is Jetpack for iOS Support, which looks to be the same for all tickets users create from this app. Is there anything we should do here?

I will clarify with HE folks.

> To create an empty line between chat messages, I tried \n\n but it seems Zendesk ignores these in their backend dashboard (they are rendered when viewing the ticket transcript in My Tickets, as shown in the video below). Zendesk's ticket dashboard might interpret markdown, so another approach could be to do what's suggested [here](https://stackoverflow.com/a/64164898/1305067) and use backslashes, but I haven't tried it yet.

I'll check more about that today. Maybe we could format each line as a different message in Zendesk

File: WordPress/WordPressTest/SupportChatBotViewModelTests.swift [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+final class SupportChatBotViewModelTests: XCTestCase {
+    private var sut: SupportChatBotViewModel!
+    private var zendeskUtils: ZendeskUtilsSpy!
+
+    override func setUpWithError() throws {
+        zendeskUtils = ZendeskUtilsSpy()
+        sut = SupportChatBotViewModel(zendeskUtils: zendeskUtils)
+    }
+
+    override func tearDownWithError() throws {
```
[the remark this answers is not in the document]
Comment (staskus) [reply WordPress/WordPressTest/SupportChatBotViewModelTests.swift:13]: Actually we shouldn't but ok

**Comment by staskus** [comment]
### Success flow

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/187f9464-b38b-484e-8f65-8b4f192afa9c

### Flow where there's no Zendesk identity + error

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/1942a75a-b61c-4001-bba2-7b3f4267ec37

**Comment by staskus** [comment]
@guarani 

I tried and failed at 2 approaches:
- Set DocsBot conversation messages as Zendesk messages for proper readability. Although `ZDKRequestProvider` has `addComment` method, it only allows to add it on behalf of the user. We wouldn't be able to add comments on behalf of a user and on behalf of a bot. Also, SDK only has `addComment` method meaning we would have to make this request for each of the messages which can be inefficient
-  Add history as a private (internal) note. Again, `addComment` doesn't allow to add a private comment, only allows to add a public comment on behalf of the user. I also tried to use `customField` and set a value to a `private note` custom field. However, in this case the message wasn't set on Zendesk

Using `Zendesk API` could potentially be an option but it would require a separate authentication and the whole setup. For now, it looked like an overreach.  Maybe @irfano  will find an option on Android.

Based on HEs request:
- I formatted history using `\n` for new-line and `>` for a space, it's not ideal but allows to at least visually see a difference between messages
- Added a `docsBot` tag

**Review by guarani — APPROVED** [review]
Thanks for adding the error handling, the Zendesk tag, and making the message formatting better. I hadn't realized the message history was an array of arrays (instead of a simple array), allowing us to separate the bot messages from the user's messages 👍 The transcript is easier to read now in Zendesk.

**Comment by irfano** [comment]
The result of my work on the task is exactly the same, @staskus. 

> SDK only has `addComment` method meaning we would have to make this request for each of the messages which can be inefficient

That's right!

> However, in this case the message wasn't set on Zendesk

Same! Adding a private note didn't work.

> Using Zendesk API could potentially be an option but it would require a separate authentication and the whole setup. For now, it looked like an overreach. Maybe @irfano will find an option on Android.

Yes, we can use [Tickets API](https://developer.zendesk.com/api-reference/ticketing/tickets/tickets/), but it would be too complex. 
The issue is that we're trying to replicate a real ZD ticket conversation by using a chat history. Maybe there are 10 messages in the chat history. For this case, we should make 5 Tickets API calls and 5 [Requests API](https://developer.zendesk.com/api-reference/ticketing/tickets/ticket-requests/) calls (what if the network disconnects? 😨).  Another drawback would be that all messages in the ticket would show the same sent-time. 
So, I think the best solution is adding the formatted chat history in a single comment from the end user, as done in this PR.

**Pull request merged**